### PR TITLE
Add recent transaction results to bank

### PR DESCRIPTION
**Pull request by garious**
Most of #1298

TODO: Needs polish
- [ ] More tests
- [x] Boot BudgetState::last_error

### CODE REVIEW

**Comment by mvines** [comment]
So far so good!  You wanna go the extra mile and plumb this all the way though rpc.rs [1]?  😅 

[1] https://github.com/solana-labs/solana/blob/master/src/rpc.rs#L91

**Comment by garious** [comment]
Sure, will do.

**Comment by aeyakovenko** [comment]
@garious BankError is pretty big and all of the metadata besides the error code can be derived by the client from the transaction itself, which they have since they have the Signature. I think it can be just an error code with no data.  We can address this in another PR though.

**Comment by garious** [comment]
I've noticed that as well. Yes, it should be split up.

File: src/bank.rs [hunk 1]
```diff
@@ -384,7 +416,9 @@ impl Bank {
         } else if BudgetState::check_id(&tx.program_id) {
             // TODO: the runtime should be checking read/write access to memory
             // we are trusting the hard coded contracts not to clobber or allocate
-            BudgetState::process_transaction(&tx, accounts)
+            if BudgetState::process_transaction(&tx, accounts).is_err() {
+                return Err(BankError::ProgramRuntimeError);
```
Comment (garious) [inline src/bank.rs]: This branch needs a unit test.

File: src/bank.rs [hunk 2]
```diff
@@ -454,6 +488,7 @@ impl Bank {
         let execution_elapsed = now.elapsed();
         let now = Instant::now();
         Self::store_accounts(&txs, &res, &loaded_accounts, &mut accounts);
+        self.update_transaction_statuses(&txs, &res);
```
Comment (garious) [inline src/bank.rs]: This side-effect needs a unit test.

File: src/budget_program.rs [hunk 3]
```diff
@@ -230,32 +230,22 @@ impl BudgetState {
...
-            let _ = Self::apply_debits_to_budget_state(tx, accounts, &instruction)
+            Self::apply_debits_to_budget_state(tx, accounts, &instruction)
                 .and_then(|_| Self::apply_credits_to_budget_state(tx, accounts, &instruction))
-                .map_err(|e| {
-                    trace!("saving error {:?}", e);
-                    Self::save_error_to_budget_state(e, accounts);
-                });
         } else {
             info!("Invalid transaction userdata: {:?}", tx.userdata);
+            Err(BudgetError::UserdataDeserializeFailure)
```
Comment (garious) [inline src/budget_program.rs]: This case is probably unreachable. If so, consider changing it to `unreachable!`.

**Comment by garious** [comment]
Merging this a little early so that @aeyakovenko can rebase on it. #1298 tracks the remaining work - aiming to complete that by Wednesday CoB.

**Pull request merged**